backend/src/org/commcare/resources/ResourceManager.java

```diff
@@ -0,0 +1,377 @@
+                // table.
+                Logger.log("resource", "Recovering global resources from recovery table");
+
+                masterTable.destroy();
+                tempTable.copyToTable(masterTable);
+
+                Logger.log("resource", "Global resources recovered. Wiping recovery table");
+                tempTable.destroy();
+            }
+        }
+
+        // Global and incoming are now in the right places. Ensure we have no
+        // uncommitted resources.
+        if (masterTable.getTableReadiness() == ResourceTable.RESOURCE_TABLE_UNCOMMITED) {
+            masterTable.rollbackCommits();
+        }
+
+        if (upgradeTable.getTableReadiness() == ResourceTable.RESOURCE_TABLE_UNCOMMITED) {
+            upgradeTable.rollbackCommits();
+        }
+
+        // If the global table needed to be recovered from the recovery table,
+        // it has. There are now two states: Either the global table is fully
+        // installed (no conflicts with the upgrade table) or it has unstaged
+        // resources to restage
+        if (masterTable.getTableReadiness() == ResourceTable.RESOURCE_TABLE_INSTALLED) {
+            Logger.log("resource", "Global table in fully installed mode. Repair complete");
+        } else if (masterTable.getTableReadiness() == ResourceTable.RESOURCE_TABLE_UNSTAGED) {
+            Logger.log("resource", "Global table needs to restage some resources");
+            masterTable.repairTable(upgradeTable);
+        }
+    }
+
+    /**
+     * Set listeners and checkers that enable communication between low-level
+     * resource installation and top-level app update/installation process.
+     *
+     * @param tableListener  allows resource table to report its progress to the
+     *                       launching process
+     * @param cancelCheckker allows resource installers to check if the
+     *                       launching process was cancelled
+     */
+    public void setUpgradeListeners(TableStateListener tableListener,
+                                    InstallCancelled cancelCheckker) {
+        masterTable.setStateListener(tableListener);
+        upgradeTable.setStateListener(tableListener);
+
+        upgradeTable.setInstallCancellationChecker(cancelCheckker);
+    }
+
+    /**
+     * @return Is the table non-empty, marked for upgrade, with all ready
+     * resources?
+     */
+    public static boolean isTableStagedForUpgrade(ResourceTable table) {
+        return (table.getTableReadiness() == ResourceTable.RESOURCE_TABLE_UPGRADE &&
+                table.isReady() &&
+                !table.isEmpty());
+    }
+
+    public boolean isUpgradeTableStaged() {
+        return isTableStagedForUpgrade(upgradeTable);
+    }
+
+    /**
+     * @return True if profile argument points to an app version that isn't
+     * any newer than the profile in the upgrade table.
+     */
+    public boolean updateIsntNewer(Resource currentProfile) {
```

Comment:
Could think about renaming this like we talked about. Even updateNotNewer sounds less awkward